core/src/main/java/org/apache/iceberg/SchemaUpdate.java

```diff
@@ -260,6 +261,64 @@ public UpdateSchema updateColumnDoc(String name, String doc) {
     return this;
   }
 
+  @Override
+  public UpdateSchema updateSchema(Schema newSchema) {
+    TypeUtil.visit(newSchema, new ApplyUpdates(schema, this));
+    return this;
+  }
+
+  private static final Joiner DOT = Joiner.on(".");
+
+  private static class ApplyUpdates extends TypeUtil.SchemaVisitor<Void> {
+    private final Schema schema;
+    private final UpdateSchema updateSchema;
+    private final Map<String, Integer> indexByName;
+
+    private ApplyUpdates(Schema schema, UpdateSchema updateSchema) {
+      this.schema = schema;
+      this.updateSchema = updateSchema;
+      this.indexByName = TypeUtil.indexByName(schema.asStruct());
+    }
+
+    @Override
+    public Void struct(Types.StructType struct, List<Void> fieldResults) {
```

Comment:
Is this not intended to handle renames or deletes?

Reply:
@rdblue .  I considered this API for supporting runtime [automatic] schema evolution. I think, in that scenario, we cannot distinguish between delete + add vs rename  reliably as the updated schema is coming from a compute engine, e.g Spark.

I think I may still be able to support field additions, type widening using IDs as you mention here - https://github.com/apache/incubator-iceberg/pull/759#discussion_r381033678.

Reply:
So do you think it is ok, If I only support field additions, updates like type widening as in this PR and not address deletes and renames ?

Reply:
There are two cases, I think. If you have field IDs for both schemas, then you can handle renames by detecting the ID matches but names do not. If you don't have field IDs then I think the most important thing is to be consistent, rather than trying to guess. I'd say go with drop + add in that case.

Reply:
I should note that in these cases, we tend to really try to get schema changes expressed as the intended changes instead of the new schema, to avoid this problem.